### PR TITLE
SY-1038 Import / Export Schematic Improvements

### PR DESCRIPTION
**Pull request by pjdotson**
# Fix Pull Request Template

## Key Information

- **Linear Issue**: [SY-1038](https://linear.app/synnaxlabs/issue/SY-1038/meaningful-error-when-importing-an-existing-schematic)

## Description

Various improvements to importing / exporting a schematic, including an export button on the visualization toolbar, fixed file drag-and-drop, and better error catching.

**IMPORTANT NOTE**: I added a `name` field to the schematic state that tracks the name of the schematic. However, migrations from previous versions will not migrate over the new names to the state, instead giving it the name "New Schematic" in the store state. 

## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added sufficient regression tests to cover the changes to CI.
- [x] I have added relevant tests to cover the changes or exposing bugs.
- [x] I have verified code coverage targets are met.

## Additional Notes
- [ ] These changes deal with concurrency.
- [x] These changes affect UI.

## Manual QA Additions

- [x] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
  with necessary manual QA steps to test my change.

## Reviewer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in readiness checklists.
- [ ] UI changes have been tested.
- [ ] style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.

### CODE REVIEW

File: console/src/layout/slice.ts [hunk 1]
```diff
@@ -100,15 +100,18 @@ interface SetAltKeyPayload {
 interface SetHaulingPayload extends Haul.DraggingState {}
 
 export interface FileHandlerProps {
-  mosaicKey: number;
   file: any;
```
Comment (Lham42) [inline console/src/layout/slice.ts:103]: does the file type have to be any - can we define an explicit type for a file?

**Review by Lham42 — CHANGES_REQUESTED** [review]

**Comment by Lham42** [comment]
is this for synnax 27?

File: console/src/schematic/hooks.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+  let name = "schematic";
+
+  return useMutation<void, Error>({
+    mutationFn: async () => {
+      const fileResponse = await open({
+        directory: false,
+        multiple: false,
+        title: `Import schematic`,
+        extenstions: ["json"],
```
Comment (emilbon99) [inline console/src/schematic/hooks.ts:87]: This prop is spelled wrong. Is it spelled wrong in the API or did you spell it wrong?

File: console/src/schematic/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import { Status, Synnax } from "@synnaxlabs/pluto";
+import { useMutation } from "@tanstack/react-query";
+import { open, save } from "@tauri-apps/plugin-dialog";
+import { readFile, writeFile } from "@tauri-apps/plugin-fs";
+import { useDispatch, useStore } from "react-redux";
+
+import { Confirm } from "@/confirm";
+import { Layout } from "@/layout";
+import { select } from "@/schematic/selectors";
+import { fileHandler } from "@/schematic/services/file";
```
Comment (emilbon99) [inline console/src/schematic/hooks.ts:20]: This import causes a dependency flow in a direction it shouldn't go. The entire point of the `services` package is to  isolated it from the root `schematic` package. Importing this dependency links the two again, potentially causing issues with circular deps in the future.

File: console/src/schematic/services/ontology.tsx [hunk 4]
```diff
@@ -155,14 +126,14 @@ const TreeContextMenu: Ontology.TreeContextMenu = (props) => {
   const del = useDelete();
   const copy = useCopy();
   const snapshot = useRangeSnapshot();
-  const download = useDownload();
+  const exprt = useExport(resources[0].name);
```
Comment (emilbon99) [inline console/src/schematic/services/ontology.tsx:129]: Typo, just call it `export`

File: console/src/schematic/toolbar/Toolbar.tsx [hunk 5]
```diff
@@ -102,7 +105,16 @@ export const Toolbar = ({ layoutKey }: ToolbarProps): ReactElement | null => {
     >
       <ToolbarHeader>
         <ToolbarTitle icon={<Icon.Schematic />}>{name}</ToolbarTitle>
-        <Tabs.Selector style={{ borderBottom: "none" }} />
+        <Align.Space direction="x" align="center" size={0}>
```
Comment (emilbon99) [inline console/src/schematic/toolbar/Toolbar.tsx:108]: Some comments on the UI:

Adding this extra button forces the "Properties" Tab to go off the page slightly, making things look strange.

There should be some breathing room between the export button and the tab selectors. Otherwise it makes it seem like the tabs and the button are elements related in functionality, which they're not. Add a divider and some space. Let it breathe!

The rounded corners on the button also look strange as the shadow of the button hits the edges of the toolbar header. Either make the button smaller so it sits well inside the toolbar, or make the edge border radius zero by using the `sharp` prop on the button.

File: console/src/schematic/hooks.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          snapshot: schematic.snapshot,
+          key: schematic.key,
+        };
+        name = schematic.name;
+      }
+      if (name == null)
+        throw new UnexpectedError("Schematic with key is missing in store state");
+      const savePath = await save({
+        defaultPath: `${name}.json`,
+        filters: [{ name: "JSON", extensions: ["json"] }],
```
Comment (emilbon99) [inline console/src/schematic/hooks.ts:49]: What do these filters do?

File: pluto/src/icon/Icon.tsx [hunk 7]
```diff
@@ -46,8 +46,12 @@ export const Icon = ({
   </div>
 );
 
-export interface CreateProps extends Omit<IconProps, "topRight"> {}
+export interface DetailProps extends Omit<IconProps, "topRight"> {}
 
-export const Create = (props: CreateProps): ReactElement => (
+export const Create = (props: DetailProps): ReactElement => (
```
Comment (emilbon99) [inline pluto/src/icon/Icon.tsx:51]: Was this rename intentional?

**Review by emilbon99 — CHANGES_REQUESTED** [review]

**Comment by Lham42** [comment]
Add a step in qa to check the migration but also note that this change is made at this specific version so we don't need to do the check again going forward

**Review by Lham42 — CHANGES_REQUESTED** [review]

File: pluto/src/icon/Icon.tsx (hunk 7 above)
Comment (emilbon99) [reply pluto/src/icon/Icon.tsx:51]: Why is it called `DetailProps`, I don't understand the naming. You could also just create a type alias so you could have `CreateProps` and `ImportProps` and just alias one to the other.

**Review by emilbon99 — APPROVED** [review]
Approved once you address my final comment

File: console/src/schematic/services/ontology.tsx (hunk 4 above)
Comment (Lham42) [reply console/src/schematic/services/ontology.tsx:137]: can we call it something else that doesn't look like a typo then

**Review by Lham42 — APPROVED** [review]
approved once you address outstanding comments
